### DISCUSSION TITLE
ME81H - Hijack an unused GPIO pin of EPS12S

### DISCUSSION THREAD

**Discussion by Kolia56** [arendst/Tasmota#18270] [Q&A] [answered]
Hello,
I replaced the original WBR3 module by an EPS12S and successfully set-up the thermostat with Tasmota 12.4.0. Except if I'm mistaken, the current hardware and hence the software does not report the status of the relay. I'm wondering if I could hijacked an unused GPIO pin. The idea would be to wire the command (with some pull-up or pull-down resistor, or voltage divider depending on the voltage that drives the relay coil) of the relay to this GPIO and then get the status of the relay. I made several tries, without any wiring so far, to see whether the status of such an supposunused GPIO could be reported. So, for instance I declared GPIO15 or 16 as LED 1. Unfortunately no LED appears in the interface. So I have a couple questions:
- Are the "unused" GPIO open for any other use when Tuya module (54) is selected?
- Are there any chances to achieve what I'm looking for?
Thank you

**Comment by sfromis** [comment]
While you can do some things with otherwise unused gpio pins, doing it for types of things which are within the scope of what the TuyaMCU chip will normally be able to control cannot coexist with the TuyaMCU setup.

Anyway, it does not make sense with that LED config, as this is an output function (a status led of Tasmota), not an input function. Typical input configs are Button and Switch, but that's the type of thing where I expect the TuyaMCU setup to override.

As I strive to just not get devices saddled with TuyaMCU, I'm not sure exactly what you can do, but I do know that in some cases you can use TuyaMCU commands to config additional details about what Tasmota will be able to set and get.

**Comment by Kolia56** [comment]
Thank you very much. If I understand well the only potential path, is to make TuyaMCU send the necessary data to ESP12S that will send it to Wifi network. I'm not familiar enough with the interaction between TuyaMCU and EPS12s to dig into that. One thing though, if it is to TuyaMCU to send the state of the relay, then it means I need to find a way to tell TuyaMCU to send over MQTT the state of the relay, isn'it ? I
Would you have in mind me some directions, either documentation or similar problems?
I have in mind a fallback solution that I'd like to avoid to set-up, that is a simple wifi switch which input is connected to the relay output, it 's a bit overkill;) I already checked the relay contacts are voltage free.

**Reply by sfromis** [reply]
Trying to keep a good distance to anything TuyaMCU, I have no specific guidance about how to battle that thing other than docs. And yes, finding a way of having the main MCU of the device report details would be useful.

The concept with TuyaMCU is this being the main chip controlling the device, and the ESP8266 (or similar) is merely used as a "Wifi modem" to their app, hence it has no direct access to device operation.

Docs: https://tasmota.github.io/docs/TuyaMCU/
Possibly useful tool: https://github.com/sillyfrog/Tasmota-Tuya-Helper/

**Comment by Kolia56** [comment]
Thank you. Yes I have in hands the very helpful Tuya helper, and nothing is reported when the relay state changes. I also read several times the Tuya MCU doc. It’s is a pity, I would risk to say it’s a kind of design hole on Tuya’s side, not reporting the relay state, that is when the heating is on or off, kind of strange. That is why I was hopping to use one free GPIO on ESP12 side. I wouldn’t be surprise to learn that when Tasmota is setup as  TuyaMCU unit, nothing else can be set up on ESP side, at least as of today. 
The solution would be then to get a thermostat that is purely based on ESP family.

**Reply by sfromis** [reply]
The Tuya design is what the Tuya app needs, not to cater to crazy people using a "Wifi chip" without Tuya firmware. I'd say that the real design hole is even having a separate MCU, instead of letting the same handle both device and Wifi.

I did not say that *nothing else* can be set up, only that it cannot conflict with TuyaMCU functions. Non-conflicting things would generally work, in my expectation. "Conflicting" does not mean only for the specific device, but also what the TuyaMCU support in Tasmota could do on some other device. 

Maybe configuring a pin as a Counter pin could sense something without conflicting with TuyaMCU. Not tried that....

**Comment by Kolia56** [comment]
> Typical input configs are Button and Switch, but that's the type of thing where I expect the TuyaMCU setup to override.

I guess you are right, I tried to declare GPIO15 as a relay or a switch, nothing appears in Tasmota GUI

However, as of for counter component I tried my luck, and yes the counter shows up in the interface and appears in the MQTT data flow. Any idea how this could be exploited?

FYI, before starting the long journey to replacing WRM3 module and understanding the TuyaMCU command set, I set up the device on the Tuya platform. Thanks to the API explorer, I analysed every details of data that is published. I did not see any relay status reporting in the data flow. So I was hoping it would change with Tasmota.
{
  "Time": "2023-03-27T08:41:04",
  "COUNTER": {
    "C1": 0
  },
  "TuyaSNS": {
    "Temperature": null,
    "TempSet": null
  },
  "Thermostat0": {
    "ThermostatModeSet": 0,
    "TempTargetSet": 18,
    "CtrDutyCycleRead": 0
  },
  "TempUnit": "C"
}

**Comment by Kolia56** [comment]
Well I think I'm on something that could be interesting. When the thermostat relay is switched off, 15" afterwards the DpId enum 36 is set to 1. Likewise when the relay is switched on, it is set to 0. There is no FnId linked to it as the herebelow screenshot tells
![image](https://user-images.githubusercontent.com/20872098/227906473-a1a841d1-d660-487c-909c-3890c49f1e85.png)

Digging into the documentation, I stumbled on tuyasend2 2,1. It revealed a new DpId 

![image](https://user-images.githubusercontent.com/20872098/227908172-af617f24-d705-4a52-8033-93bec1858f3c.png)

It turns out that this DpId 2 drives the heater relay. However, when the unit set the relay on or off, this DpId is not updated. I noticed though that DpId is synced with DpId 2

So the question is: how to send over MQTT the state of DpId 36?

**Reply by sfromis** [reply]
Probably involving using the TuyaMCU command to assign that dpId to a Tasmota [fnId](https://tasmota.github.io/docs/TuyaMCU/#fnid).

**Reply by Kolia56** [reply]
I did that already. See current configuration

`14:54:39.448 MQT: stat/tasmota_0E7070/RESULT = {"TuyaMCU":[{"fnId":11,"dpId":1},{"fnId":42,"dpId":2},{"fnId":72,"dpId":16},{"fnId":71,"dpId":24},{"fnId":61,"dpId":36}]}`

But nothing appears in the interface. I've got the feeling that I have to learn how to set-up a rule that will publish the state of this enum whenever there is a change. Any idea of what should be the rule to set-up?

**Reply by sfromis** [reply]
As repeatedly alluded to, my TuyaMCU knowledge is deliberately slim, much lower than the page I linked.

**Comment by Kolia56** [comment, marked as answer]
For the records, chased data is embedded in some "TuyaReceived" stream. Tuya MCU helper is really a must have in that kind of situation. 
Two rules are set in order to get the thermostat relay state pushed to MQTT, hence to HA:
`rule1 on TuyaReceived#Data=55AA03070005240400010037 DO publish stat/tasmota_0E7070/RELAY 1 endon on TuyaReceived#Data=55AA03070005240400010138 DO publish stat/tasmota_0E7070/RELAY 0 endon`

**Reply by ortegafernando** [reply]
will it be possible to "integrate" this to tasmota tuya mcu comms? Is it a "new command" from the tuya MCU?

**Reply by Kolia56** [reply]
I'd really like to help you but I'm afraid I don't understand what you are looking for

**Comment by g6094199** [comment]
for the ME81H.16 the seems to be an dpld for the relay:

```
DpType4Id36":1,"36":{"DpId":36,"DpIdType":4,"DpIdData":"01"}}}   --> Relay off
DpType4Id36":0,"36":{"DpId":36,"DpIdType":4,"DpIdData":"00"}}}   --> Relay on 
```

taken from this very usefull topic at [IObroker](https://forum.iobroker.net/topic/47322/raumthermostat-me81h-31-wifi-tuya-tasmota-flashen/74)

and generally you can use a gpio for detecting the relay. you just have to find the line where the relay gets its power (should be 5V), add an voltage devider of, lets say, 2x 20k ohm to ground and feed the middle of the devider (2,5V) to a gpio. now you can detect on/off.

and there are some gpios unused eg. the 2 pins next to the uart0. is use them for i2c to add an additional sht4x for humidity

**Reply by sfromis** [reply]
Here's the pinout: https://tasmota.github.io/docs/Pinouts/#esp-12s
Then it's up to you to check what's already in use.

Beware that the TuyaMCU setup precludes many types of features also available from a TuyaMCU, superseding similar base Tasmota features, as the TuyaMCU setup is **not** meant for you to add more stuff do. Still, some things may work out.

**Comment by Kolia56** [comment]
> for the ME81H.16 the seems to be an dpld for the relay:
> 
> ```
> DpType4Id36":1,"36":{"DpId":36,"DpIdType":4,"DpIdData":"01"}}}   --> Relay off
> DpType4Id36":0,"36":{"DpId":36,"DpIdType":4,"DpIdData":"00"}}}   --> Relay on 
> ```
> 
Thank you. I confirm your statement. With the rule I set, this is another way to get the relay state reported to MQTT. 

> and generally you can use a gpio for detecting the relay. you just have to find the line where the relay gets its power (should be 5V), add an voltage devider of, lets say, 2x 20k ohm to ground and feed the middle of the devider (2,5V) to a gpio. now you can detect on/off.
> 
So to me, no need to link a GPIO to the relay command, and possibly geopardize the unit. BTW this was my first idea that I abandonned.

**Comment by Kolia56** [comment]
> Beware that the TuyaMCU setup precludes many types of features also available from a TuyaMCU, superseding similar base Tasmota features, as the TuyaMCU setup is **not** meant for you to add more stuff do. Still, some things may work out.

Thank you. BTW I'm struggling to try to get the Tasmota Thermostat feature working without success so far. You may review what I already wrote on this topic is this [post](https://github.com/arendst/Tasmota/discussions/19836#discussioncomment-7916826), especially the section related to PI feature. Do you have an idea?

**Reply by sfromis** [reply]
No, not using the thermostat feature myself, I have no how-to suggestions.

But as a starting point, I'd not have an expectation of it working with at TuyaMCU infested device (unless experimenting or having clear info otherwise), as this this leaves Tasmota without any direct control, instead reduced to act similar to what can be done from a "remote control" app. While this is way different from standard operating condition for any feature, the TuyaMCU driver does try to take over some basic Tasmota operations to mimic those via TuyaMCU. Details are not clear to me, as I've made sure to not get into the troubles with TuyaMCU-infested devices.

**Reply by Kolia56** [reply]
This is the right way to describe those devices as infected. When I started the journey with ME81H I was miles away from my current understanding. There are a nice piece of hardware loaded with a very poor software. The presence of TuyaMCU makes it very difficult to implement any other solution. I’m wondering whether EPS Home would help on this matter.
I’ve been searching the Net for any device like this one based on ESP device. Are you aware of such devices?
